[webkitcorepy] Add mutually exclusive groups to TaskPool

JonWBedard opened the pull request:
#### 03d9366f146b13e52b370ee15d905a7da16b6c01
<pre>
[webkitcorepy] Add mutually exclusive groups to TaskPool
<a href="https://bugs.webkit.org/show_bug.cgi?id=273695">https://bugs.webkit.org/show_bug.cgi?id=273695</a>
<a href="https://rdar.apple.com/127503729">rdar://127503729</a>

Reviewed by Sam Sneddon.

Add the concept of mutually exclusive groups to TaskPool, which will
allow for types of tests to be grouped in a single processes with other
tests of their own type.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
(_State.__init__): Include mutually exclusive groups supported by worker.
(_State.__call__): Print names of mutually exclusive groups supported by worker.
(_Queue): Added.
(_Process.main): Attempt to fetch tasks from any groups registered to this worker
before fetching tasks from the shared queue.
(TaskPool.__init__): Allow caller to define mutually exclusive groups.
(TaskPool.__enter__): Distribute mutually exclusive groups amongst workers, with each
worker taking ownership over an equal number of groups.
(TaskPool.do): Allow caller to specify a group send task to.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
(action): Add option to log which worker a task was run on.
(TaskPoolUnittest.test_mutually_exclusive_group):
(TaskPoolUnittest.test_mutually_exclusive_groups):
(TaskPoolUnittest.test_invalid_group):

Canonical link: <a href="https://commits.webkit.org/278787@main">https://commits.webkit.org/278787@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b1d8dee537f0a8cbddbb89ef594bfea2ca8e66cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51571 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30881 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3924 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/54838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2264 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53874 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37217 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1945 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/54838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/53670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/28518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44481 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/54838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/51420 "Found 1 webkitpy python3 test failure: webkit.messages_unittest.GeneratedFileContentsTest.test_message_argument_description") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/1742 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1832 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56430 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/26691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/51590 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/27930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44549 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/28824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7518 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/27665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
```diff
@@ -170,3 +173,75 @@ def test_teardown_arguments(self):
+
+        def test_mutually_exclusive_group(self):
+            with OutputCapture(level=logging.INFO) as captured:
+                with TaskPool(workers=4, mutually_exclusive_groups=dict(group=lambda c, **kwargs: c in ['a', 'b', 'c', 'd'])) as pool:
```
As I understand it (and I only really understand it from the tests), `mutually_exclusive_groups` is a dict from group name (`str`) to a callable that is truthful if it is a member of that group? And the callable gets passed the arguments the action receives?

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
```diff
@@ -264,7 +296,7 @@ def handler(cls, value, _):
 
     @classmethod
-    def main(cls, name, loglevel, setup, setupargs, setupkwargs, queue, teardown, teardownargs, teardownkwargs):
+    def main(cls, name, loglevel, setup, setupargs, setupkwargs, queue, teardown, teardownargs, teardownkwargs, mutually_exclusive_groups):
```
It's slightly surprising to me that this is a dict here, versus a list everywhere else. Not sure it makes sense to change anything however?

JonWBedard replied:
I'll change this to `mutuals_exclusive_group_queues`, it's an internal argument anyways.

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
```diff
@@ -297,7 +330,22 @@ def main(cls, name, loglevel, setup, setupargs, setupkwargs, queue, teardown, te
+                            break
+                    if not task:
+                        task = queue.receive()
+                    # Since queue.receive is blocking, we should check group_queues one final time before breaking the work look
```
```suggestion
                    # Since queue.receive is blocking, we should check group_queues one final time before breaking the work loop
```

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
```diff
@@ -170,3 +173,93 @@ def test_teardown_arguments(self):
+            )
+
+        def test_mutually_exclusive_groups(self):
+            self.maxDiff = None
```
Let's not land this.

gsnedders reviewed (commented):
My main comment here is the typo in the comment; this basically all looks fine at this point.

JonWBedard committed:
[webkitcorepy] Add mutually exclusive groups to TaskPool
https://bugs.webkit.org/show_bug.cgi?id=273695
rdar://127503729

Reviewed by Sam Sneddon.

Add the concept of mutually exclusive groups to TaskPool, which will
allow for types of tests to be grouped in a single processes with other
tests of their own type.

* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
(_State.__init__): Include mutually exclusive groups supported by worker.
(_State.__call__): Print names of mutually exclusive groups supported by worker.
(_Queue): Added.
(_Process.main): Attempt to fetch tasks from any groups registered to this worker
before fetching tasks from the shared queue.
(TaskPool.__init__): Allow caller to define mutually exclusive groups.
(TaskPool.__enter__): Distribute mutually exclusive groups amongst workers, with each
worker taking ownership over an equal number of groups.
(TaskPool.do): Allow caller to specify a group send task to.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/task_pool_unittest.py:
(action): Add option to log which worker a task was run on.
(TaskPoolUnittest.test_mutually_exclusive_group):
(TaskPoolUnittest.test_mutually_exclusive_groups):
(TaskPoolUnittest.test_invalid_group):

Canonical link: https://commits.webkit.org/278787@main
